extras/dht/dht.c

```diff
@@ -0,0 +1,128 @@
+ *  @expected_pin_state     high (true) or low (false) pin state
+ *  @counter                pointer to external uint8_t for tallying the duration waited for the pin state
+*/
+
+bool waitForPinState(uint8_t pin, uint8_t interval, uint8_t timeout, bool expected_pin_state, uint8_t * counter) {
+
+    for (*counter = 0; *counter < timeout; *counter+=interval) {
+        if (gpio_read(pin) == expected_pin_state) return true;
+        sdk_os_delay_us(interval);
+    }
+
+    return false;
+}
+
+/*  
+ *  
+ *  
+ *  @pin                    the selected GPIO pin
+ *  @humidity               pointer to external int8_t to store resulting humidity value
+ *  @temperature            pointer to external int8_t to store resulting temperature value
+ *  @status_message         pointer to external 100-character string to store status messages
+*/
+
+bool readHumidityAndTemperature(int8_t pin, int8_t * humidity, int8_t * temperature, char ** status_message) {
+    int8_t data[40] = {0};
+    int8_t result[5] = {0};
+    uint8_t i = 0;
+
+    uint8_t init_phase_duration = 0;
+    uint8_t duration_data_low = 0;
+    uint8_t duration_data_high = 0;
+
+    gpio_enable(pin, GPIO_OUT_OPEN_DRAIN);
+    gpio_set_pullup(pin, false, false);
+
+    taskENTER_CRITICAL();
+
+    // Phase 'A' pulling signal low to initiate read sequence
+    gpio_write(pin, 0);
+    sdk_os_delay_us(20000);
+    gpio_write(pin, 1);
+
+    // Step through Phase 'B' at 2us intervals, 40us max
+    if (waitForPinState(pin, 2, 40, false, &init_phase_duration)) {
+        // Step through Phase 'C ' at 2us intervals, 88us max
+        if (waitForPinState(pin, 2, 88, true, &init_phase_duration)) {
+            // Step through Phase 'D' at 2us intervals, 88us max
+            if (waitForPinState(pin, 2, 88, false, &init_phase_duration)) {
+
+                // Read in each of the 40 bits of data...
+                for (i = 0; i < 40; i++) {
+                    if (waitForPinState(pin, 2, 60, true, &duration_data_low)) {
+                        if (waitForPinState(pin, 2, 75, false, &duration_data_high)) {
+                            data[i] = duration_data_high > duration_data_low;
+                        }
+                    }
+                }
+
+                taskEXIT_CRITICAL();
+
+                for (i = 0; i < 40; i++) {
+                    // Read each bit into 'result' byte array...
+                    result[i/8] <<= 1;
+                    result[i/8] |= data[i];
+                }
+
+                if (result[4] == ((result[0] + result[1] + result[2] + result[3]) & 0xFF)) {
+                    // Data valid, checksum succeeded...
+                    *humidity = result[0];
+                    *temperature = result[2];
+                    *status_message = "Success...";
+                    return true;
+                } else {
+                    *status_message = "Checksum failed, invalid data received from sensor...";

```

Comment:
I wasn't really suggesting switching between returning a code vs strings, as that would get really ugly with C's static typing and make the API a pain to work with on the application side too.  I was mainly just suggesting you could add debugging `printf` statements in the driver code which could be enabled at compile time if the user wants to know more about what's going on.  What I usually do is something like the following (at the top of the code):

```c
#ifndef DEBUG_DHT
#define DEBUG_DHT 0
#endif

#if DEBUG_DHT
#define debug(fmt, ...) printf("%s" format "\n", "dht: ", ## __VA_ARGS__);
#else
#define debug(fmt, ...) /* (do nothing) */
#endif
```
..and then you can just use `debug("debug message!")`, etc, throughout your code..  if the project is built with `-DDEBUG_DHT=1`, the debug messages will be included and will be printed to stdout when it executes.  If not (the default), the debug statements and their string arguments, etc, won't even be included in the compiled output.

Actually, to be honest, in this case the more I think about it, the more I think you probably don't need to return even a status code, really.. you could just change all the `*status_message = "..."` lines to `debug("...")` instead.  I can't really come up with any scenarios where it would be useful/important for the caller to know whether the failure was in (for example) phase B vs phase C of the transaction..  that's probably only useful for diagnostic/debugging level stuff..